File: curso-c-completo/modulo-1-fundamentos/ejemplos/04_entrada_salida.c
/*
 * ============================================================================
 * PROGRAMA: Entrada y Salida en C
 * DESCRIPCIÓN: Aprende a leer datos del usuario con scanf()
 * NIVEL: Principiante
 * ============================================================================
 */

#include <stdio.h>

int main() {

    // ========== LECTURA DE UN ENTERO ==========

    printf("========== LECTURA DE DATOS ==========\n\n");

    int edad;

    printf("¿Cuál es tu edad? ");

    /*
     * scanf() lee datos del teclado
     * &edad es la "dirección de memoria" de la variable edad
     * El & (ampersand) es OBLIGATORIO con scanf
     * %d indica que esperamos un número entero
     */
    scanf("%d", &edad);

    printf("Tienes %d años.\n\n", edad);

    // ========== LECTURA DE UN DECIMAL ==========

    float altura;

    printf("¿Cuál es tu altura en metros? ");
    scanf("%f", &altura);

    printf("Mides %.2f metros.\n\n", altura);

    // ========== LECTURA DE UN CARÁCTER ==========

    char inicial;

    printf("¿Cuál es la inicial de tu nombre? ");

    /*
     * IMPORTANTE: scanf deja un 'Enter' en el buffer
     * Usamos un espacio antes de %c para ignorar espacios en blanco
     */
    scanf(" %c", &inicial);

    printf("Tu inicial es: %c\n\n", inicial);

    // ========== LECTURA MÚLTIPLE ==========

    int dia, mes, anio;

    printf("Ingresa tu fecha de nacimiento (dd mm aaaa): ");
    // Podemos leer múltiples valores en un solo scanf
    scanf("%d %d %d", &dia, &mes, &anio);

    printf("Naciste el %d/%d/%d\n\n", dia, mes, anio);

    // ========== EJEMPLO PRÁCTICO: CALCULADORA DE IMC ==========

    printf("========== CALCULADORA DE IMC ==========\n\n");

    float peso, altura_m, imc;

    printf("Ingresa tu peso en kg: ");
    scanf("%f", &peso);

    printf("Ingresa tu altura en metros: ");
    scanf("%f", &altura_m);

    // IMC = peso / (altura * altura)
    imc = peso / (altura_m * altura_m);

    printf("\n--- RESULTADOS ---\n");
    printf("Peso: %.2f kg\n", peso);
    printf("Altura: %.2f m\n", altura_m);
    printf("Tu IMC es: %.2f\n", imc);

    // Interpretación del IMC
    printf("\nInterpretación:\n");
    printf("  Bajo peso: < 18.5\n");
    printf("  Normal: 18.5 - 24.9\n");
    printf("  Sobrepeso: 25.0 - 29.9\n");
    printf("  Obesidad: >= 30.0\n");

    // ========== FORMATO DE SALIDA AVANZADO ==========

    printf("\n========== FORMATO DE SALIDA ==========\n\n");

    int numero = 42;
    float pi = 3.14159;

    // Ancho de campo
    printf("Número con ancho 10: [%10d]\n", numero);
    printf("Número con ancho 5:  [%5d]\n", numero);

    // Alineación a la izquierda (con -)
    printf("Alineado izq:        [%-10d]\n", numero);

    // Rellenar con ceros
    printf("Rellenado con 0:     [%010d]\n", numero);

    // Decimales
    printf("\nPi con diferentes decimales:\n");
    printf("  2 decimales: %.2f\n", pi);
    printf("  4 decimales: %.4f\n", pi);
    printf("  8 decimales: %.8f\n", pi);

    // Notación científica
    printf("\nNotación científica: %e\n", pi);

    return 0;
}

/*
 * CONCEPTOS CLAVE:
 *
 * printf() - SALIDA:
 *   - Imprime datos en la pantalla
 *   - Usa especificadores de formato (%d, %f, %c, etc.)
 *   - No necesita & antes de las variables
 *
 * scanf() - ENTRADA:
 *   - Lee datos del teclado
 *   - Usa los mismos especificadores de formato
 *   - REQUIERE & antes de las variables (excepto con arrays/strings)
 *   - Espera a que el usuario presione Enter
 *
 * ESPECIFICADORES DE FORMATO COMUNES:
 *   %d o %i  - int
 *   %u       - unsigned int
 *   %ld      - long
 *   %f       - float o double (entrada)
 *   %lf      - double (entrada, recomendado)
 *   %c       - char
 *   %s       - string (cadena de texto)
 *
 * MODIFICADORES DE FORMATO:
 *   %10d     - Ancho mínimo de 10 caracteres
 *   %-10d    - Alineado a la izquierda
 *   %010d    - Rellena con ceros
 *   %.2f     - 2 decimales
 *   %10.2f   - Ancho 10, 2 decimales
 *
 * CARACTERES ESPECIALES:
 *   \n       - Nueva línea
 *   \t       - Tabulación
 *   \\       - Barra invertida
 *   \"       - Comillas dobles
 *
 * ERRORES COMUNES:
 *
 * 1. Olvidar el & en scanf:
 *    ✗ scanf("%d", edad);      // ERROR
 *    ✓ scanf("%d", &edad);     // CORRECTO
 *
 * 2. Usar %f para double en scanf:
 *    ✗ scanf("%f", &mi_double);   // Puede causar problemas
 *    ✓ scanf("%lf", &mi_double);  // CORRECTO
 *
 * 3. No limpiar el buffer antes de leer char:
 *    ✗ scanf("%c", &letra);    // Puede leer el Enter anterior
 *    ✓ scanf(" %c", &letra);   // El espacio limpia el buffer
 *
 * EJERCICIOS:
 *
 * 1. Programa que pida dos números y muestre:
 *    - Suma
 *    - Resta
 *    - Multiplicación
 *    - División
 *
 * 2. Conversor de temperaturas:
 *    - Lee temperatura en Celsius
 *    - Muestra equivalente en Fahrenheit
 *    - Fórmula: F = C * 9/5 + 32
 *
 * 3. Calculadora de área de círculo:
 *    - Lee el radio
 *    - Calcula área (π * radio²)
 *    - Calcula perímetro (2 * π * radio)
 *
 * 4. Programa que lea 3 notas y calcule el promedio
 *
 * 5. Conversor de monedas:
 *    - Lee cantidad en dólares
 *    - Lee tasa de cambio
 *    - Muestra equivalente en otra moneda
 */

File: curso-fortran/modulo-1-fundamentos/04_entrada_salida.f90
! ============================================================================
! Programa: Entrada y Salida de Datos
! Descripción: Aprende a leer datos del usuario y mostrar resultados
! ============================================================================

program entrada_salida
    implicit none

    ! Declaración de variables
    character(len=50) :: nombre
    integer :: edad, año_nacimiento
    real :: altura, peso, imc
    character(len=30) :: ciudad
    real :: celsius, fahrenheit

    ! ========================================================================
    ! ENTRADA BÁSICA DE DATOS
    ! ========================================================================

    print *, '======================================='
    print *, 'PROGRAMA DE ENTRADA Y SALIDA'
    print *, '======================================='
    print *

    ! Solicitar el nombre
    print *, 'Por favor, ingresa tu nombre:'
    read(*,*) nombre  ! Lee una línea del teclado

    ! Solicitar la edad
    print *, 'Ingresa tu edad:'
    read(*,*) edad

    ! Solicitar la altura
    print *, 'Ingresa tu altura en metros (ej: 1.75):'
    read(*,*) altura

    ! Solicitar el peso
    print *, 'Ingresa tu peso en kilogramos:'
    read(*,*) peso

    ! Solicitar la ciudad
    print *, 'Ingresa tu ciudad:'
    read(*,*) ciudad

    ! ========================================================================
    ! PROCESAMIENTO DE DATOS
    ! ========================================================================

    ! Calcular año de nacimiento (aproximado)
    año_nacimiento = 2025 - edad

    ! Calcular IMC (Índice de Masa Corporal)
    imc = peso / (altura * altura)

    ! ========================================================================
    ! SALIDA FORMATEADA DE DATOS
    ! ========================================================================

    print *
    print *, '======================================='
    print *, 'RESUMEN DE TU INFORMACIÓN'
    print *, '======================================='
    print *
    print *, 'Nombre:', trim(nombre)  ! trim() elimina espacios en blanco
    print *, 'Edad:', edad, 'años'
    print *, 'Año de nacimiento (aprox):', año_nacimiento
    print *, 'Altura:', altura, 'm'
    print *, 'Peso:', peso, 'kg'
    print *, 'Ciudad:', trim(ciudad)
    print *, 'IMC:', imc
    print *

    ! Interpretación del IMC
    if (imc < 18.5) then
        print *, 'Tu IMC indica: Bajo peso'
    else if (imc >= 18.5 .and. imc < 25.0) then
        print *, 'Tu IMC indica: Peso normal'
    else if (imc >= 25.0 .and. imc < 30.0) then
        print *, 'Tu IMC indica: Sobrepeso'
    else
        print *, 'Tu IMC indica: Obesidad'
    end if

    print *
    print *, '======================================='

end program entrada_salida

! ============================================================================
! CÓMO FUNCIONA LA ENTRADA/SALIDA:
!
! SALIDA (mostrar datos):
!   print *, 'mensaje'           - Salida con formato libre
!   write(*,*) 'mensaje'         - Equivalente a print
!
! ENTRADA (leer datos):
!   read(*,*) variable           - Lee del teclado
!
! El primer * en read/write se refiere a la unidad de E/S:
!   * = entrada/salida estándar (teclado/pantalla)
!   5 = entrada estándar
!   6 = salida estándar
!
! El segundo * se refiere al formato:
!   * = formato libre
!   También puedes especificar formatos personalizados
!
! COMPILAR Y EJECUTAR:
!   gfortran -o entrada_salida 04_entrada_salida.f90
!   ./entrada_salida
! ============================================================================

! EJERCICIOS:
! 1. Crea un programa que pida dos números y muestre su suma, resta,
!    multiplicación y división
! 2. Haz un programa que convierta temperaturas de Celsius a Fahrenheit
!    Fórmula: F = (C × 9/5) + 32
! 3. Programa que calcule el área de un triángulo pidiendo base y altura
! 4. Programa que pida el radio de un círculo y calcule área y perímetro
